### PR TITLE
Publish TB0T only when it's valid and allow opt-out

### PR DESCRIPTION
**Pull request by zhen-zen**

### CODE REVIEW

File: Sensors/SMCBatteryManager/SMCBatteryManager.cpp [hunk 1]
```diff
@@ -61,6 +61,8 @@ bool SMCBatteryManager::start(IOService *provider) {
 	
 	DBGLOG("bmgr", "AppleSMC is available now");
 
+	BatteryManager::getShared()->start();
```
Comment (vit9696) [inline Sensors/SMCBatteryManager/SMCBatteryManager.cpp:64]: Ugh, are you sure of this change?

Comment (zhen-zen) [reply Sensors/SMCBatteryManager/SMCBatteryManager.cpp:64]: I moved it before SMC key submission so it's possible to update configuration and values, which is used to decide whether `TB0T`s should present. The ACPI polls should work well on its own, or did I missing something?

Comment (vit9696) [reply Sensors/SMCBatteryManager/SMCBatteryManager.cpp:64]: Right, that is probably ok, just a little worried of the organisation…

**Comment by vit9696** [comment]
@usr-sse2 @lvs1974 this looks good to me, please check this out. I want it to be merged in December release.

**Comment by lvs1974** [comment]
It looks fine for me as well.

File: Sensors/SMCBatteryManager/BatteryManagerState.hpp [hunk 2]
```diff
@@ -55,10 +55,10 @@ struct BatteryInfo {
 		uint8_t chargeLevel {0};
 		uint32_t designCapacityWarning {0};
 		uint32_t designCapacityLow {0};
-		uint16_t temperatureRaw {0};
+		uint16_t temperatureKelvin {2931};
```
Comment (usr-sse2) [inline Sensors/SMCBatteryManager/BatteryManagerState.hpp]: If it's in decikelvins, then name it `temperatureDecikelvin`

File: Sensors/SMCBatteryManager/SMCBatteryManager.cpp [hunk 3]
```diff
@@ -85,10 +87,12 @@ bool SMCBatteryManager::start(IOService *provider) {
 		VirtualSMCAPI::addKey(KeyB0RM(i), vsmcPlugin.data, VirtualSMCAPI::valueWithUint16(2000, new B0RM(i), SMC_KEY_ATTRIBUTE_PRIVATE_WRITE|SMC_KEY_ATTRIBUTE_WRITE|SMC_KEY_ATTRIBUTE_READ));
 		VirtualSMCAPI::addKey(KeyB0St(i), vsmcPlugin.data, VirtualSMCAPI::valueWithData(nullptr, 2, SmcKeyTypeHex, new B0St(i), SMC_KEY_ATTRIBUTE_PRIVATE_WRITE|SMC_KEY_ATTRIBUTE_WRITE|SMC_KEY_ATTRIBUTE_READ));
 		VirtualSMCAPI::addKey(KeyB0TF(i), vsmcPlugin.data, VirtualSMCAPI::valueWithUint16(0, new B0TF(i)));
-		VirtualSMCAPI::addKey(KeyTB0T(i+1), vsmcPlugin.data, VirtualSMCAPI::valueWithSp(0, SmcKeyTypeSp78, new TB0T(i)));
+		if (BatteryManager::getShared()->state.btInfo[i].state.temperature != -20) {
```
Comment (usr-sse2) [inline Sensors/SMCBatteryManager/SMCBatteryManager.cpp]: Why is -20°C an invalid temperature? If TjMAX is e.g. 99°C, then temperatures can be monitored from -28°C to +99°C. Also floating point comparisons are not reliable, although this particular case is OK (small integers are precisely represented as `double`, and the value `-20` is implicitly converted to `double` and becomes exactly the same as the initial value).

Comment (zhen-zen) [reply Sensors/SMCBatteryManager/SMCBatteryManager.cpp]: I picked it from [Smart Battery Data Specification](http://sbs-forum.org/specs/sbdat110.pdf) `5.1.9 Temperature()`, typically the range is -20 to +75 Celsius. Can you suggest some values or approaches?

Comment (usr-sse2) [reply Sensors/SMCBatteryManager/SMCBatteryManager.cpp]: Typical values are not allowed range. The same paragraph of the spec states that the range is 0°K to 6553.5°K. You can either compare `temperatureKelvin` to 0 and check `BSSTemperatureSMBusOnly` again here, or add a flag.

Comment (zhen-zen) [reply Sensors/SMCBatteryManager/SMCBatteryManager.cpp]: My bad. I thought `state` is inaccessible in `getBatteryInfo`, but it turned out it's `updateRealTimeStatus` that can only access `state`. Almost forgot `state` resides under `BatteryInfo` and tried to restrict it in the same scope. Thank you.

**Comment by Lorys89** [comment]
I no longer work temperatures after this commit
laptop ice lake

**Comment by Lorys89** [comment]
before
Schermata 2020-12-07 alle 21 26 24


now
Schermata 2020-12-07 alle 21 28 13

**Comment by zhen-zen** [comment]
Please check if `SMCProcessor` is loaded.

File: Sensors/SMCBatteryManager/SMCBatteryManager.cpp (hunk 3 above)
Comment (Lorys89) [reply Sensors/SMCBatteryManager/SMCBatteryManager.cpp]: > Why is -20°C an invalid temperature? If TjMAX is e.g. 99°C, then temperatures can be monitored from -28°C to +99°C. Also floating point comparisons are not reliable, although this particular case is OK (small integers are precisely represented as `double`, and the value `-20` is implicitly converted to `double` and becomes exactly the same as the initial value).

after upgrading to kext release, i don't read the temperatures anymore, I did in my fork the revert and I compiled kext, and now read back the temperatures, what breaks the reading in these changes?

Comment (zhen-zen) [reply Sensors/SMCBatteryManager/SMCBatteryManager.cpp]: The change here is outdated. Please use bugtracker for issues other than DM and comments under commits. And you haven't attached troubleshooting files, specifically your ACPI implementation and the log from `SMCBatteryManager` and `SMCProcessor`.

A suspect for missing temperature could be the non-zero check for temperature, resulting from slow firmware initialization. That exception will be logged as well. But I still have no idea why if will affect `SMCProcessor`.

**Pull request merged**